# Проектная работа: Анализ сервиса для чтения книг по подписке.

Цель исследования: 

Проанализировать базу данных сервиса для чтения книг по подписке и сформулировать ценностное предложение для нового продукта.

# Введение

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Наша задача проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**План иследования:**

Шаг 1. Загрузка данных.

Шаг 2. Исследование таблиц.

Шаг 3. Анализ таблиц.

Шаг 4. Вывод

## Шаг 1. Загрузка данных.

In [1]:
pip install psycopg2-binary 

Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd
import sqlalchemy as sa
from sqlalchemy import create_engine 

In [3]:
db_config = {
'user': 'praktikum_student', 
'pwd': 'Sdf4$2;d-d30pp', 
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, 
'db': 'data-analyst-final-project-db' 
}

In [4]:
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [5]:
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})

In [6]:
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
    with engine.connect() as con:
        return pd.read_sql(sql=sa.text(query), con = con)

## Исследование таблиц:

### Таблица `books`

In [7]:
books =( '''
SELECT 
    *
FROM books
    LIMIT 5
''')    
get_sql_data(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [8]:
books =( '''
SELECT 
    COUNT(book_id),
    COUNT(DISTINCT(book_id)) distinct,
    COUNT(DISTINCT(title)) distinct_book
FROM books
    LIMIT 5
''')    
get_sql_data(books)

count  distinct  distinct_book
0   1000      1000            999

Данные в таблице `books` представлены следующими столбцами: `book_id` идентификатор книги, `author_id` идентификатор автора,  `title` название книги, `num_pages` количество страниц, `publication_date` дата публикации книги, `publisher_id` идентификатор издателя. В таблице 1000 строк. Уникальных названий книг 999

### Таблица `authors`

In [9]:
authors =( '''
SELECT 
    *
FROM authors
    LIMIT 5
''')    
get_sql_data(authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [10]:
authors =( '''
SELECT 
    COUNT(author_id),
    COUNT(DISTINCT(author_id)) distinct,
    COUNT(DISTINCT(author)) distinct_author
FROM authors
    LIMIT 5
''')    
get_sql_data(authors)

count  distinct  distinct_author
0    636       636              636

Данные в таблице `authors` представлены следующими столбцами: `author_id` идентификатор автора,  `author` имя автора.  В таблице 636 строк. Уникальных авторов 636

### Таблица `publishers`

In [11]:
publishers =( '''
SELECT 
    *
FROM publishers
    LIMIT 5
''')    
get_sql_data(publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [12]:
publishers =( '''
SELECT 
    COUNT(publisher_id),
    COUNT(DISTINCT(publisher_id)) distinct,
    COUNT(DISTINCT(publisher)) distinct_publisher
FROM publishers
    LIMIT 5
''')    
get_sql_data(publishers)

count  distinct  distinct_publisher
0    340       340                 340

Данные в таблице `publishers` представлены следующими столбцами: `publisher_id` идентификатор издателя, `publisher` название издательства. В таблице 340 строк. Уникальных издательств 340

### Таблица `ratings`

In [13]:
ratings =( '''
SELECT 
    *
FROM ratings
    LIMIT 5
''')    
get_sql_data(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [14]:
ratings =( '''
SELECT 
    COUNT(rating_id),
    COUNT(DISTINCT(rating_id)) distinct,
    COUNT(DISTINCT(username)) distinct
FROM ratings
    LIMIT 5
''')    
get_sql_data(ratings)

count  distinct  distinct
0   6456      6456       160

Данные в таблице `ratings` представлены следующими столбцами: `rating_id` идентификатор оценки, `book_id` идентификатор книги,   `username` имя пользователя, оставившего оценку, `rating` оценка книги.  В таблице 6456 строк. Уникальных оценевающих пользователей 160.

### Таблица `reviews`

In [15]:
reviews =( '''
SELECT 
    *
FROM reviews
    LIMIT 5
''')    
get_sql_data(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [16]:
reviews =( '''
SELECT 
    COUNT(review_id),
    COUNT(DISTINCT(review_id)) distinct,
    COUNT(DISTINCT(username)) distinct
FROM reviews
    LIMIT 5
''')    
get_sql_data(reviews)

count  distinct  distinct
0   2793      2793       160

Данные в таблице `reviews` представлены следующими столбцами: `review_id` идентификатор обзора, `book_id` идентификатор книги,   `username` имя пользователя написавшего обзор, `text` текст обзора.  В таблице 2793 строк. Уникальных критикующих пользователей 160.

## Детализированное иследование

### Анализ даты публикации книги.
Посчитайте, сколько книг вышло после 1 января 2000 года

In [17]:
zad_1 =( '''

SELECT 
    COUNT(book_id) 
FROM books
    WHERE CAST(publication_date AS date) > '2000-01-01'

''')    
get_sql_data(zad_1)

count
0    819

Начиная с первого января 2000 года, на сервиве по подписке книг, было опубликованно 819 книг 

### Анализ количества обзоров и средней оценки.

Для каждой книги посчитайте количество обзоров и среднюю оценку.

In [18]:
zad_2 =( '''

WITH 

rat AS (
SELECT 
    a.author book,
    AVG(rt.rating) avg_rating
FROM books AS b 
JOIN authors a ON b.author_id=a.author_id
LEFT JOIN ratings AS rt ON b.book_id=rt.book_id
    GROUP BY a.author
    ORDER BY a.author
),

rev AS(
SELECT 
    a.author book,
    COUNT(rv.review_id) count_review
FROM books AS b 
JOIN authors a ON b.author_id=a.author_id
LEFT JOIN reviews AS rv ON b.book_id=rv.book_id
    GROUP BY a.author
    ORDER BY a.author
)

SELECT 
    rat.book,
    rev.count_review,
    ROUND(rat.avg_rating, 2) avg_rating
FROM rat 
JOIN rev ON rat.book=rev.book
    LIMIT 5
''')    
get_sql_data(zad_2)

book  count_review  avg_rating
0                      A.S. Byatt             3        4.00
1  Aesop/Laura Harris/Laura Gibbs             2        3.75
2                 Agatha Christie            28        4.28
3                   Alan Brennert             4        3.00
4        Alan Moore/David   Lloyd             3        4.22

Пользовательские рейтенги книг как и количество проставленных оценок отличаются у разных книг.

In [19]:
zad_2 =( '''

WITH 

rat AS (
SELECT 
    b.book_id book,
    AVG(rt.rating) avg_rating
FROM books AS b 
LEFT JOIN ratings AS rt ON b.book_id=rt.book_id
    GROUP BY b.book_id
    ORDER BY b.book_id
),

rev AS(
SELECT 
    b.book_id book,
    COUNT(rv.review_id) count_review
FROM books AS b 
LEFT JOIN reviews AS rv ON b.book_id=rv.book_id
    GROUP BY b.book_id
    ORDER BY b.book_id
),

alls AS(
SELECT 
    rat.book,
    rev.count_review,
    ROUND(rat.avg_rating, 2) avg_rating
FROM rat 
JOIN rev ON rat.book=rev.book
)
 
SELECT 
    ROUND(AVG(count_review), 2) avg_review,    
    MAX(count_review) max_review,
    MIN(count_review) min_review,
    ROUND(AVG(avg_rating), 2) avg_rating,
    MAX(avg_rating) max_rating,
    MIN(avg_rating) min_rating
FROM
    alls

    
   
''')    
get_sql_data(zad_2)

avg_review  max_review  min_review  avg_rating  max_rating  min_rating
0        2.79           7           0         3.9         5.0         1.5

Среднее количество просмотров рамно 2,79, максимальное количество просмотров 7,
минимальное 0. Средний рейтинг книги равен 3,9 максимальный 5, минимальный 1,5

### Анализ издательств

Определите издательство, которое выпустило наибольшее число книг толще
50 страниц — так вы исключите из анализа брошюры;

In [20]:
zad_3 =( '''

SELECT 
    COUNT(b.book_id),
    p.publisher_id,
    publisher
FROM books b
LEFT JOIN publishers p ON b.publisher_id=p.publisher_id
    WHERE num_pages > 50
    GROUP BY p.publisher_id
    ORDER BY COUNT(b.book_id) DESC
LIMIT 1

''')    
get_sql_data(zad_3)

count  publisher_id      publisher
0     42           212  Penguin Books

`Penguin Books` Это издательство выпустило 42 книги объемом более 50 страниц, что является максимальным количеством таких книг 

### Анализ авторов

Определите автора с самой высокой средней оценкой книг — учитывайте
только книги с 50 и более оценками;

In [21]:
zad_4 =( '''
SELECT 
    a.author_id
FROM author a
LEFT JOIN books b ON a.author_id=b.author_id
WHERE b.book_id IN (
    SELECT 
        b.book_id
    FROM books b
    JOIN authors a ON b.author_id=a.author_id
    LEFT JOIN ratings r ON b.book_id=r.book_id
        GROUP BY b.book_id
        HAVING COUNT(r.username) >= 50
        ORDER BY AVG(r.rating) DESC, COUNT(r.username) DESC
   )
LIMIT 1
''')    
get_sql_data(zad_4)

author_id
0        236

In [22]:
zad_4 =( '''

SELECT 
    a.author,
    ROUND(AVG(rating), 6) avg_rating
FROM authors a
LEFT JOIN books b ON a.author_id=b.author_id
LEFT JOIN ratings r ON r.book_id=b.book_id
WHERE b.book_id IN (
    SELECT 
        b.book_id
    FROM books b
    JOIN authors a ON b.author_id=a.author_id
    LEFT JOIN ratings r ON b.book_id=r.book_id
        GROUP BY b.book_id
        HAVING COUNT(r.username) >= 50
        ORDER BY  COUNT(r.username) DESC)
GROUP BY a.author    
ORDER BY AVG(rating) desc
LIMIT 1


''')    
get_sql_data(zad_4)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

Максимальный средний рейтинг, среди книг с пятидесятью и более оценками, равен 4,29 и он принаджежит слудующему автору `J.K. Rowling/Mary GrandPré`. 

### Анализ обзоров

Посчитайте среднее количество обзоров от пользователей, которые поставили
больше 48 оценок.

In [23]:
zad_5 =( '''

SELECT 
    AVG(count_reviews) avg_reviews
FROM (
    SELECT 
        COUNT(review_id) count_reviews,
        username
    FROM reviews
        WHERE username IN (
            SELECT 
                username
            FROM ratings
                GROUP BY username
                HAVING COUNT(rating_id) > 48
            )
        GROUP BY username    
) AS f


''')    
get_sql_data(zad_5)

avg_reviews
0         24.0

In [24]:
zad_5 =( '''

SELECT 
    COUNT(count_reviews) count_reviews
FROM (
    SELECT 
        COUNT(review_id) count_reviews,
        username
    FROM reviews
        WHERE username IN (
            SELECT 
                username
            FROM ratings
                GROUP BY username
                HAVING COUNT(rating_id) > 48
            )
        GROUP BY username    
) AS f


''')    
get_sql_data(zad_5)

count_reviews
0             13

Средннее количество обзоров пользователей, которые поставили более 48 оценок, равно 24 обзора. Общее количество таких пользователей 13

## Вывод

Данные в таблице `books` представлены следующими столбцами: `book_id` идентификатор книги, `author_id` идентификатор автора,  `title` название книги, `num_pages` количество страниц, `publication_date` дата публикации книги, `publisher_id` идентификатор издателя. В таблице 1000 строк. Уникальных названий книг 999

Данные в таблице `authors` представлены следующими столбцами: `author_id` идентификатор автора,  `author` имя автора.  В таблице 636 строк. Уникальных авторов 636

Данные в таблице `publishers` представлены следующими столбцами: `publisher_id` идентификатор издателя, `publisher` название издательства. В таблице 340 строк. Уникальных издательств 340

Данные в таблице `ratings` представлены следующими столбцами: `rating_id` идентификатор оценки, `book_id` идентификатор книги,   `username` имя пользователя, оставившего оценку, `rating` оценка книги.  В таблице 6456 строк. Уникальных оценевающих пользователей 160.

Данные в таблице `reviews` представлены следующими столбцами: `review_id` идентификатор обзора, `book_id` идентификатор книги,   `username` имя пользователя написавшего обзор, `text` текст обзора.  В таблице 2793 строк. Уникальных критикующих пользователей 160.

Начиная с первого января 2000 года, на сервиве по подписке книг, было опубликованно 819 книг

Пользовательские рейтенги книг как и количество проставленных оценок отличаются у разных книг.

Среднее количество обзоров ровно 2,79, максимальное количество просмотров 7,
минимальное 0. Средний рейтинг книги равен 3,9 максимальный 5, минимальный 1,5

`Penguin Books` Это издательство выпустило 42 книги объемом более 50 страниц, что является максимальным количеством таких книг 

Максимальный средний рейтинг, среди книг с пятидесятью и более оценками, равен 4,29 и он принаджежит слудующему автору `J.K. Rowling/Mary GrandPré`. 


Средннее количество обзоров пользователей, которые поставили более 48 оценок, равно 24 обзора.

### Итоговый вывод

Куплен маленький сервис для чтения книг по подписке,  в сервисе представлено всего 1000 книг 81% из которых современные.  На сервисе публикуются 636 авторов из 340  издательств. Рейтинги и обзоры на сервисе отличаются, они "не накручены", но активная аудитория малеленькая, всего из 160 пользователей 13 активных.

### Рекомендации
- необходимо увеличивать количество книг на сервисе.
- необходимо привлекать как авторов так и издательства.
- Необходимо увеличивать как саму аудиторию читателей, так и ее активность.
